tests/optimization/bfgs.cpp

```diff
@@ -0,0 +1,172 @@
+/*
+   Copyright (c) 2009-2016, Ryan H. Lewis
+   All rights reserved.
+
+   This file is part of Elemental and is under the BSD 2-Clause License, 
+   which can be found in the LICENSE file in the root directory, or at 
+   http://opensource.org/licenses/BSD-2-Clause
+*/
+#include <utility>
+#include <El.hpp>
+#include <El/optimization/bfgs.hpp>
+#define CATCH_CONFIG_RUNNER
+#include <catch.hpp>
+
+using namespace El;
+
+template< typename T>
+std::pair< DistMatrix<T>, T>
+LogisticRegression( const DistMatrix<T>& X, const DistMatrix<T>& y, const T lambda=1){
+  const std::function< T(const DistMatrix<T>&)>
+  logistic_function = [&](const DistMatrix<T>& theta){
+     auto nrm = El::Norm(theta);
+     El::DistMatrix<T> y(theta);
+     El::DistMatrix<T> r(X.Height(), 1);
+     El::Gemv(El::NORMAL, T(-1), X, theta, T(0), r);
+     for( auto i = 0; i < y.Height(); ++i){ r.Set(i, 0, y.Get(i,0));  }
+
+     T v = lambda*nrm*nrm;
+     for(auto i = 0; i < y.Height(); ++i){ v += El::Log(T(1) + El::Exp(-1*y.Get(i,0)*r.Get(i,0))); }
```

Comment:
This is going to be monstrously slow due to (1) being linear runtime due to every process performing linear work, and (2) due to there being `2*y.Height()` broadcasts within the `Get` calls. It would be better to locally perform the sums and then to reduce at the end with a single `mpi::AllReduce`.